### ISSUE TITLE
IfcPatch Recipe Optimise error message

### ISSUE THREAD

**Issue by maxfb87** [IfcOpenShell/IfcOpenShell#1393]
Hi,
i got this message when i try to run the IfcPatch Optimize
![error](https://user-images.githubusercontent.com/79401028/111539575-5066d180-876e-11eb-8421-f157a8624b88.JPG)
I only set the input .ifc file and the output file (new one)

**Comment by Moult** [comment]
Put in "[]" in the arguments field.

**Comment by maxfb87** [comment]
Ok, i tried and i got this error
![image](https://user-images.githubusercontent.com/79401028/111699342-490bfa80-8838-11eb-9765-b269724fd01c.png)
After that, i have installed the toposort module in the Blender python and now it seems to work...

Btw with a 14MB file, after 10minutes of working, Blender doesn't respond...i had forced the quit...

**Comment by Moult** [comment]
@maxfb87 agree, it's not a very optimised optimise module :) See #1043. It hasn't been fixed yet. Ping @johltn .

Maybe the RecycleNonRootedElements recipe will help? It's another approach which does the same thing which may or may not be less painful.

**Comment by maxfb87** [comment]
So there should be a recipe for optimize the optimize recipe :-) @Moult 

Btw, the optimize process doesn't finish at all, neither after 30min of working...
Is there a way to see the command that creates the problem? I mean a log or something...

I will try also the RecicleNonRootedElement and i will tell you.
Thanks

**Comment by maxfb87** [comment]
@Moult so, i tried also the RecicleNonRootedElement and it created the "clean" file correctly in about 20sec.
File dimension passed from about 14 MB  to about 7 MB.

But i opened the "clean" file in ifc++ and now there are new errors that there wasn't before; there are also some minor graphical issues in representation.

**Comment by johltn** [comment]
Hi all, I'll have a look at it in the next days.

**Comment by Moult** [comment]
Yeah I'm not too confident in the recycle recipe. The optimise recipe is in theory more correct and thorough, though it is slow. The recycle recipe is simpler but makes some broad assumptions and isn't battle tested.

I think I need to create a new recipe that has more finegrained assumptions and less broad. Even if the optimise recipe is fast, it still optimises everything in the graph. Some of these relationships actually have meaning (e.g. reuse of psets, layer sets, structural boundary conditions) and others don't (e.g. reuse of coordinates, rgbs), so maybe a fresh attempt is needed.

**Comment by aothms** [comment]
> Some of these relationships actually have meaning

The schema explicitly states that unless it's a rooted element (and then it won't be recycled since it has a different GlobalId) there is no semantic difference between shared/deduplicated and duplicate resources. (forgot where it is, but I have been pointed to it in the past)

**Comment by Moult** [comment]
@aothms yeah I think you're right. I guess I'm a little suspicious since I have found some conflicting evidence. Here's a few that I recently came across in the past few weeks.

`IfcMaterialLayerSetUsage` states (emphasis mine):

> The IfcMaterialLayerSetUsage is always assigned to an **individual** occurrence object (and only to relevant subtypes of IfcElement). The IfcMaterialLayerSet, referenced by ForLayerSet, **can however be shared** among several occurrence objects.

This is not explicit, but it is certainly suggestive that there are some rules around reuse.

I have also come across some structural analysis models where IfcBoundaryCondition have no `Name` attribute, but are clearly grouped with meaning in their reuse. If the optimiser ran on this, it would mangle the groups. I guess that's the fault of the user who isn't naming their boundary conditions, but...

Then there is the fact that IfcMaterial isn't rooted, whereas it probably should be.

Anyway - you're right. The proper solution would be to optimise the optimiser recipe :) I take back my proposal to revisit it and create a new recipe.

**Comment by aothms** [comment]
I think you're correct. There can indeed constraints on the inverse cardinality that should be taken into account in this recipe (not sure how hard it would be, or we can set some entity types probably that we don't optimise.). I thought such a constraint was indeed present in IfcMaterialLayerSetUsage, but now I can't find it (@hlg didn't you also mentioned that? I can't find any cardinality constraint for the material definitions).

An additional complication that we probably don't take into account sufficiently is that we only hash the forward attributes, but consider the following example where there are two identical extrusions but one with a (inversely) assigned style. We shouldn't fold `#1` and `#2` into the same instance, but I think it'll happen. Sadly get_info() doesn't do inverses yet, but it would also mean that every instance references the entire graph, not very scalable.

    #1=extrusion(#x,#y,#z)
    #2=extrusion(#x,#y,#z)
    #3=styled_item(#2,(#w),$)

**Comment by hlg** [comment]
I don't know if it affects this issue, but there is a lower bound of 1 on the cardinality of the inverse `IfcMaterialUsageDefinition.AssociatedTo` (invese of `IfcRelAssociatesMaterial.RelatingMaterial`. Thus `IfcMaterialLayerSetUsage` can not exist without being associated to a material. That's in IFC4 only.

**Comment by aothms** [comment]
Thanks @hlg that's indeed what I was looking for, but wrongly remembered. It's a lower bound and not an upper bound. Thanks for the reminder

**Comment by Moult** [comment]
Bump :) @johltn did you manage to take a peek at it?

**Comment by maxfb87** [comment]
@hlg @aothms @Moult @johltn i think i can close this issue. Is it ok or there is yet something to do?

**Comment by aothms** [comment]
@maxfb87 sure, if nobody looked into it in the past 3y I guess nobody will in the near future.
